rep-2012.rst

```diff
@@ -0,0 +1,466 @@
+
+Client ID and sequence number
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Both the client ID and sequence number are provided by the ROS middleware [5]_.
+They can be accessed from ``rcl`` [6]_ when taking a request or response for a service or client respectively.
+
+Example
+^^^^^^^
+
+For example, consider a service ``example_interfaces/srv/AddTwoInts`` defined as follows:
+
+.. code-block::
+
+   int64 a
+   int64 b
+   ---
+   int64 sum
+
+The following service event message definitions are generated when building the ``example_interfaces`` package (comments elided for brevity):
+
+:example_interfaces/msg/AddTwoInts_RequestEvent:
+
+.. code-block::
+
+   uint8 CLIENT_SENT      = 0
+   uint8 SERVICE_RECEIVED = 1
+
+   uint8 request_type
+
+   builtin_interfaces/msg/Time stamp
+
+   unique_identifier_msgs/msg/UUID client_id
+
+   int64 sequence_number
+
+   example_interfaces/srv/AddTwoInts_Request request
+
+:example_interfaces/msg/AddTwoInts_ResponseEvent:
+
+.. code-block::
+
+   uint8 SERVICE_SENT     = 0
+   uint8 CLIENT_RECEIVED  = 1
+
+   uint8 response_type
+
+   builtin_interfaces/msg/Time stamp
+
+   unique_identifier_msgs/msg/UUID client_id
+
+   int64 sequence_number
+
+   example_interfaces/srv/AddTwoInts_Response response
+
+The definition for ``example_interfaces/srv/AddTwoInts_Request`` is,
+
+.. code-block::
+
+   int64 a
+   int64 b
+
+And the definitions for ``example_interfaces/srv/AddTwoInts_Reponse`` is,
+
+.. code-block::
+
+   int64 sum
+
+Configuration
+-------------
+
+Configuration of service introspection features will be done through ROS parameters.
+Each node may offer the following four Boolean parameters:
+
+:publish_client_events:
+  If true, then client "request sent" and "response received" events will be published for all clients created by this node.
+:publish_service_events:
+  If true, then service "request received" and "response sent" events will be published for all services created by this node.
+:client_event_content:
+  If true, then client event messages will include the request and response content.
+  Only applies if ``publish_client_events`` is ``true``.
+:service_event_content:
+  If true, then service event messages will include the request and response content.
+  Only applies if ``publish_service_events`` is ``true``.
+
+By default, all parameters shall be ``false`` so users do not pay for a feature they do not plan to use.
+Furthermore, node authors may opt-in by default or disable the service introspection feature altogether as they see fit.
+
+As an example, users can enable service introspection for all services by providing the following parameters YAML file to nodes [7]_:
+
+.. code-block:: yaml
+
+   /**:
+      ros__parameters:
+         publish_service_events: true
+         # Also include request and response content in event messages
+         service_event_content: true
```

Comment:
This would imply that we either publish out the service/client content two times each for a total of four times, or not at all. For larger services this can be significant -- there may be merit in enabling more granular control.

Reply:
To clarify, in this example we are only enabling service side events (not from the client), for a total of two events: _request received_ (with request content) and _response sent_ (with response content).